### PR TITLE
Register function race

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This test fixes a race condition. Profiling relies upon core worker. When a worker is shutdown the core worker is unset first. The import thread could still be importing a new function which requires profiling though. 

## Related issue number

Unclear if this was the original bug in #7073, but that did reproduce it. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/import_thread.py
Comment (ericl) [inline python/ray/import_thread.py:145]: Can you add a unit test?

Comment (ericl) [reply python/ray/import_thread.py:145]:
```
import ray

ray.init(num_cpus=1)


jobs = []
for _ in range(8):

    @ray.remote(max_calls=1)
    def foo():
        return

    jobs.append(foo.remote())
ray.get(jobs)
```

This snippet seems to do it.

Comment (wuisawesome) [reply python/ray/import_thread.py:145]: We discussed this, but just writing it here. Not adding a unit test because it's not worth it. This bug is actually quite minor because only causes the worker to die after shutdown is called (therefore it doesn't meaningfully effect the program's behavior). Not worth parsing stderr for the right error message.

**Review by mehrdadn — COMMENTED** [review]
LGTM, but I'm not familiar enough with these code paths to approve—I'll defer to others' reviews.

File: src/ray/core_worker/core_worker.cc
Comment (ericl) [inline src/ray/core_worker/core_worker.cc:151]: Can we keep the check? It is better than segfaulting.

Comment (edoakes) [reply src/ray/core_worker/core_worker.cc:151]: +1, wouldn't be surprised if this gets introduced again in the future

File: python/ray/import_thread.py
Comment (edoakes) [reply python/ray/import_thread.py:145]: @wuisawesome is there another way to fix this aside from just disabling the profiling? We basically just need to make sure import thread stops before the core worker is destructed, right? Not a huge deal but would be better to keep this in the timeline.

**Comment by wuisawesome** [comment]
Ok, it's unclear if this is worth merging as a stop gap measure. In the long term, it would be nice to keep the profiling and we should make sure we terminate all python processes during the shutdown procedure. @edoakes I think this it's your call if you think we should merge this.

**Review by ericl — APPROVED** [review]

File: python/ray/import_thread.py
Comment (ericl) [reply python/ray/import_thread.py:145]: Stopping the thread from C++ seems pretty hard, I don't think it's really worth it for now.

**Comment by ericl** [comment]
@wuisawesome make sure to check failing tests:
```
=================================== FAILURES ===================================
______________________________ test_profiling_api ______________________________

E               ray.test_utils.RayTestTimeoutException: Timed out while waiting for information in profile table. Missing events: {'register_remote_function'}.
```

**Comment by edoakes** [comment]
@wuisawesome I think it's worth merging for now. @rkooo567 is going to do some work to improve the existing profiling code so maybe he can take a second look at this.

**Comment by rkooo567** [comment]
@wuisawesome Will you merge this PR?

**Comment by kfstorm** [comment]
I think a better way to fix it is to provide a callback for core worker to invoke during shutting down. In the callback, we stop the background Python threads.

**Comment by wuisawesome** [comment]
@kfstorm  I agree. 

@rkooo567 Are you sure this is causing the failure?

**Comment by rkooo567** [comment]
You mean the tagged issue? I am sure the issue is `create_profile` is called after core worker is destructed. I also agree @kfstorm's solution is the right approach to solve this (I think we discussed this through slack before). I am okay to merge this PR as a temporary solution unless someone is willing to spend some time on fundamental fix in a few weeks.

**Comment by kfstorm** [comment]
It's OK to have a temporary fix. Can we keep the RAY_CHECK? Seems unrelated to this PR.

**Comment by wuisawesome** [comment]
This isn't a release blocker so I think we should just do it properly when someone has the bandwidth, or if/when the profiling code is overhauled.

**Comment by ericl** [comment]
@wuisawesome please make sure your PRs are passing travis tests once they are approved. Then, you can ping for merge.

**Comment by wuisawesome** [comment]
@ericl I'm just going to close this because I don't think we should merge it.

**Comment by ericl** [comment]
Why don't we merge it, it's a pretty bad error.

**Pull request merged**
